apps/system/js/sync_state_machine.js

```diff
@@ -0,0 +1,147 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+/*
+
+       +---------------Success ---------------+
+       |                                      |
+   +---+----+                                 |
+   |Enabling+---Error---+                     |
+   +---+----+           |                     |
+       ^              +-v---+                 |
+       |   +-Disable--+Error+-------------+   |
+       |   |          +--+--+             |   |
+     Enable|             ^             Enable |
+       |   |             |                |   |
+       |   |             |                |   |
+  +----+---v-+           |            +---v---v+
+  | Disabled |         Error +------->+Enabled +---+
+  +----^-----+           |   |        +---+----+   |
+       |                 |   |            |        |
+       |                 | Success       Sync      |
+       |                 |   |            |        |
+       |             +---+---+            |        |
+       |             |Syncing<------------+        |
+       |             +-------+                     |
+       |                                           |
+       +--------------Disable  --------------------+
+
+*/
+
+'use strict';
+
+(function(exports) {
+
+  var SyncStateMachine = {
+    current: 'disabled'
+  };
+
+  /**
+   * Object representing the possible events and the associated states.
+   * {
+   *   eventName: {
+   *     from: to,
+   *     from: to
+   *     ...
+   *   },
+   *   ...
+   * }
+   */
+  const events = {
+    enable: {
+      'disabled': 'enabling',
+      'error': 'enabled'
+    },
+    disable: {
+      'enabled': 'disabled',
+      'error': 'disabled'
+    },
+    sync: {'enabled': 'syncing'},
+    success: {
+      'enabling': 'enabled',
+      'syncing' : 'enabled'
+    },
+    error: {
+      'enabling': 'error',
+      'syncing' : 'error'
+    }
+  };
+
+  /**
+   * Creates the state transition function per each event.
+   *
+   * For each state transition we trigger an event with name 'onSTATE' where
+   * STATE is the name of the state where the machine transitioned to.
+   * This event gets the name of the event that caused the transition and
+   * the previous state as arguments along with the additional arguments
+   * specific to the event causing the transition.
+   *
+   * Example of an event 'A' causing a transition from state 'X' to state 'Y':
+   *
+   * SyncStateMachine.A('dummyArg');
+   *
+   * SyncStateMachine.onY = function(event, from) {
```

Comment:
This works, but maybe it would be clearer to call A  an "action", and use the word "onY event" to mean entering into state Y. Because it's customary to think of 'on...' functions as "event" handlers, so then it's easy to confuse A and Y. Also, I think it's nice if all information is on the event objects, and not in extra arguments, but maybe that's a matter of test. So maybe something like:
````js
SyncStateMachine.onY = function(event) {
  alert(event.action); // 'A'
  alert(event.from); // 'X'
  alert(event.actionArguments); // '{'0': 'dummyArgs'}'
};
````
Just an idea. :)